Oscar matrix

GDeFranceschi opened the pull request:
The file MatGrp.jl contains the definition of the types `MatrixGroup` and `MatrixGroupElem`, that is not a subtype of `GAPGroupElem{MatrixGroup}` (this allows to store the vector of generators), plus some functions for groups that need to be treated apart for matrix groups.
The file iso_oscar_gap.jl contains the isomorphisms between GAP and AbstractAlgebra objects (fields and matrices). At the moment, it works just for fields of type `FqNmodFiniteField`, but it can be extended.
Do not merge! There is still some work going on, and also tests and documentation to be written.

GDeFranceschi committed:
File MatGrp.jl

GDeFranceschi committed:
Basic methods

GDeFranceschi committed:
Generators and index

Commit:
Membership

Commit:
ring_iso types

GDeFranceschi committed:
New getproperty + some functions

GDeFranceschi committed:
Type redefinition

GDeFranceschi committed:
New type redefinition

GDeFranceschi committed:
New functionalities + iso Nemo->GAP fields

GDeFranceschi committed:
File iso_oscar_gap.jl

GDeFranceschi committed:
Minor changes

GDeFranceschi committed:
Starting replacing iso fields in MatrixGroup type

GDeFranceschi committed:
Isomorphism works

GDeFranceschi committed:
Constructors

GDeFranceschi committed:
Little change subgroups

GDeFranceschi committed:
More constructors, +tests and docstrings

GDeFranceschi committed:
Little changes + tests

GDeFranceschi committed:
Adding type AbstractMatrixGroupElem

GDeFranceschi committed:
Remove GAPGroupElem{T} + some tests

ThomasBreuer commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,805 @@
+   
+   return y
+end
+
```
Conversions to GAP will in general need more context. In the case of finite field elements that can be constructed using `Z(q)`, the situation is easy, but for example algebraic extensions in characteristic zero will require that the field on the GAP side gets created just once, and then this GAP object must be used throughout the whole session; this means that the GAP field (or some object that describes the conversion) has to be cached somewhere, such that it can be fetched when the Oscar field is given; perhaps the same is useful in the other direction as well.

ThomasBreuer commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,805 @@
+      return z^d
+   end
+end
+
```
This conversion assumes that the two fields `Fgap` and `F` are defined w.r.t. the same polynomial, and that this polynomial is primitive.
According to the documentation of `FlintFiniteField` (`Nemo.GF` is not documented), this function uses the Conway polynomial when this  is available, otherwise a random sparse irreducible polynomial. (If the polynomial is not primitive, `gen(F)` cannot be used like in this function.)
In order to make the conversion safe, one could throw an error when the two defining polynomials are different.

GDeFranceschi replied:
Yes, in fact the actual method to convert GAP fields to Oscar fields and viceversa is the one used in the file `iso_oscar_gap.jl`. I think the lines from 717 to the end of the file `MatGrp.jl` can be discarded.

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -34,11 +34,9 @@ export
 _gap_filter(::Type{PermGroup}) = GAP.Globals.IsPermGroup
 _gap_filter(::Type{PcGroup}) = GAP.Globals.IsPcGroup
 _gap_filter(::Type{FPGroup}) = GAP.Globals.IsFpGroup
-_gap_filter(::Type{MatrixGroup}) = GAP.Globals.IsMatrixGroup
```
This shouldn't be removed, though, rather it should be adapted to the new type. (Perhaps it should be *moved* to another place, but not completely replace

GDeFranceschi replied:
The function `_gap_filter` was used in other kind of groups, where there always is the underlying GAP object. This makes a code line like

> function quaternion_group(::Type{T}, n::Int) where T <: GAPGroup 
> return T(GAP.Globals.QuaternionGroup(_gap_filter(T), n))
> end

working. In the case of matrix groups, this wouldn't work, because the constructor for matrix groups works in a different way, so the code would be something like
> function quaternion_group(::Type{T}, n::Int) where T <: GAPGroup 
> if T<: MatrixGroup return ...
> else return T(GAP.Globals.QuaternionGroup(_gap_filter(T), n))
> end

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -34,11 +34,9 @@ export
 # TODO: matrix group handling usually is more complex: there usually
 # is another extra argument then to specify the base field
-#_gap_filter(::Type{MatrixGroup}) = GAP.Globals.IsMatrixGroup
 
```
I think the comment above the removed line is quite relevant to this PR .. ;-) Perhaps it should be adjusted or removed?

Commit:
Tests + a couple of bugs

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+
+   function MatrixGroup(m::Int64, F::S) where S<:Ring
```
```suggestion
   function MatrixGroup(m::Int64, F::Ring) 
```

rfourquet replied:
Also, I don't think we support 32-bits platforms, but is there a reason not to use `Int` instead of `Int64` here? (and elsewhere in this PR).

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+   mat_iso::GenMatIso
+
+   function MatrixGroup(m::Int64, F::S) where S<:Ring
+      K = elem_type(S)
```
```suggestion
      K = elem_type(F)
```

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+      return r
+   end
+
+   function MatrixGroup(m::Int64, F::S, G_gap::GapObj) where S<:Ring
```
I would suggest re-using the previous constructor instead of copy-pasting, and maybe not having it as an internal constructor then. E.g.
```julia
function MatrixGroup(m::Int64, F::Ring, G_gap::GapObj)
    M = MatrixGroup(m, F)
    M.X = G_gap
    return M
end
```
Similarly for the third constructor (and for `MatrixGroupElem`). It makes it easier to understand what the differences are between the constructors.

GDeFranceschi replied:
So, do I move the lines

> function MatrixGroup(m::Int64, F::Ring, G_gap::GapObj)
    M = MatrixGroup(m, F)
    M.X = G_gap
    return M
end

outside the struct?

rfourquet replied:
Yes. It's generally better to minimize the number of internal constructor (they are meant to enforce internal consistency of the objects; so one or two such constructors are usually enough, and other constructors can rely on them without duplicating the initialization logic).

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+      z = new{RE, MatElem{RE}}()
+      z.parent = G
+      z.elm = x
+      return z
```
Out of curiosity, why not simply `return new{RE, MatElem{RE}}(G, x)`  ?

GDeFranceschi replied:
I just followed the convention I found on the file `src/HeckeTypes.jl`. I didn't expect the shorter version works, but it does, so I can replace it.

rfourquet replied:
Note that calling `new` with arguments can sometimes have a small performance advantage, by not initializing non-bitstype fields twice (a pointer internally). But here it probably doesn't matter at all.

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+
+function Base.show(io::IO, x::MatrixGroupElem)          #TODO: is this correct?
+   if isdefined(x, :elm)
+      display(x.elm)
```
This seems weird to use `display` here, but at least this expression must use `io`.

GDeFranceschi replied:
The difference between `display` and `print(io, ...` is that in the second case the `n x n` entries of a matrix are displayed on a single row. That's why I prefer the first option. But I think this depends on the Oscar version, because I got different results working with different laptops.

rfourquet replied:
Maybe you are looking for `show(stdout, "text/plain", x)` ?

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+   if sym === :mat_iso || sym === :ring_iso
+      if isdefined(G,sym) return getfield(G,sym) end
+      fm = gen_mat_iso(G.deg, G.ring)
+      setfield!(G,:ring_iso,fm.fr)
```
Why not `G.ring_iso = fm.fr` ? `setfield!` is intended to be used when you have the name of the field in a variable, whose values is not known statically.

GDeFranceschi replied:
I did not use `G.ring_iso = fm.fr` because I was not sure to create infinite recursions. If it works, would it be sensible to change it?

rfourquet replied:
Oh I didn't think about that... is `setproperty!` overloaded somewhere?

rfourquet replied:
But yes, if possible, `G.ring_iso  =...` is much better than `setfield(G, :ring_iso, ...)`.

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+            setfield!(G,:ring_iso,fm.fr)
+            setfield!(G,:mat_iso,fm)
+         end
+         if G.descr==:GL setfield!(G,:X,GAP.Globals.GL(G.deg,G.ring_iso.codomain))
```
It would be great to not duplicate this long chain of `if-elsefi` here, and for example to factor it out in a function.

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+########################################################################
+
+# this saves the value of x.X
+# x_gap = x.X if this is already known, x_gap = Nothing otherwise
```
It's surprising to use `Nothing` the type as a sentinel here, why not `nothing` ?

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+   x_gap = GAP.Globals.Random(G.X)
+   x_oscar = G.mat_iso(x_gap)
+   x = MatrixGroupElem(G,x_oscar,x_gap)
+   return x
```
```suggestion
   return MatrixGroupElem(G,x_oscar,x_gap)
```

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+########################################################################
+
+function Base.show(io::IO, x::GroupConjClass{T,S}) where T <: MatrixGroup where S <: MatrixGroupElem
+  show(x.repr)
```
`io` must be inserted in the call, same below.

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,808 @@
+end
+
+function conjugacy_classes_maximal_subgroups(G::MatrixGroup)
+  L = GAP.gap_to_julia(Vector{GapObj},GAP.Globals.ConjugacyClassesMaximalSubgroups(G.X))
```
```suggestion
   L = GAP.gap_to_julia(Vector{GapObj},GAP.Globals.ConjugacyClassesMaximalSubgroups(G.X))
```

GDeFranceschi committed:
Suggestions from PR #173

Update src/Groups/MatGrp.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

Update src/Groups/MatGrp.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

Update src/Groups/MatGrp.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

Update src/Groups/MatGrp.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

Suggestions from #PR173 + Tests

GDeFranceschi committed:
Parenting sort out + tests

GDeFranceschi committed:
Minor changes

Minor changes

Little changes

Minor change

Update src/Groups/MatGrp.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

Minor changes

Minor changes

rfourquet commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+
+"""
+    det(x::MatrixGroupElem)
+Return the determinant of `x`.
```
We just cleaned-up our docstrings. We now try to have an empty line between the signature and the docstring:
```suggestion
    det(x::MatrixGroupElem)
    
Return the determinant of `x`.
```

thofma commented:
Rebase would be good to get CI running.

GDeFranceschi committed:
Bugs in tests + GAPGroupElem{DirectProductGroup} & similar

rfourquet commented on test/Groups/directproducts.jl:
```diff
@@ -6,7 +6,7 @@
    @test order(G)==order(S)*order(C)
    @test exponent(G)==lcm(exponent(S),exponent(C))
-   @test typeof(rand(G))==GAPGroupElem{DirectProductGroup}
+   @test typeof(rand(G))<:GAPGroupElem{DirectProductGroup}
```
```suggestion
   @test rand(G) isa GAPGroupElem{DirectProductGroup}
```

GDeFranceschi committed:
Minor changes

GDeFranceschi committed:
gens(G,i) --> gen(G,i)

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+  L=GAP.Globals.Iterator(G.X)
+  if GAP.Globals.IsDoneIterator(L)
+    return nothing
+  end
```
Groups are never empty, so how could this ever happen?
```suggestion
  @assert ! GAP.Globals.IsDoneIterator(L)
```

GDeFranceschi replied:
I asked myself this questions too, but I just copied the procedure I used for generic groups too. Is it the case to correct in the file `GAPGroups.jl` too?

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+end
+
+
+# this saves the value of x.X
```
"this" = `lies_in`?  I don't understand what this sentence, "this saves the value of x.X", is supposed to mean. Could you please elaborate?

GDeFranceschi replied:
The embedding functions such as `G(x)` (for `G` group and `x` element/matrix) return a `MatrixGroupElem` value, but the underlying GAP object `x.X`, if required, is computed in the function `lies_in` (called by the embedding function). Hence, to avoid recomputing `x.X`, the function `lies_in` returns the value `x.X` as second output in order to be used in the embedding function. But I don't even know if just delete that line, since it sound to me a quite obvious fact.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+
+# this saves the value of x.X
+# x_gap = x.X if this is already known, x_gap = nothing otherwise
+function lies_in(x::MatElem, G::MatrixGroup, x_gap)
```
OK it seems `lies_in` is somehow an alternative method to `in` which tries to avoid work for membership testing in some cases, but adds unnecessary extra work in other cases.

Overall, I don't see the point of this function:
1. we could instead teach GAP to perform these better \in tests for GL and SL (if it doesn't already); this would then benefit many more calculations.

2. to implement GL, SL, ... for "new" rings not supported by GAP, we don't have an underlying GAP object and thus a lot of functionality in this file won't work anyway; yes, `lies_in` will work, but what's the point? Not too many applications create `GL(n,R)` and then test for membership in there... In fact I think that's a super rare case that perhaps happens in artificial examples or teaching, but not in research. Just to clarify: of course it regularly happens that a computation spits out a group on which we perform membership tests; and that group then may turn out to be GL, SL, ... However, we don't *know* that unless the group was explicitly created that way, which is what I call "rare". So we are optimizing a rare case but pessimize the common case. 

3. if we really want to deal with groups GAP cannot handle right now, we'll either have to teach GAP about them in some (limited) fashion, or else add another type of matrix groups which are not backed by GAP (and then possibly redesign the type hierarchy to let that fit in nicely). But this simply hack ain't it...

4. In the end, one needs full group recognition to deal with matrix groups, period

GDeFranceschi replied:
1. Yes, that's a good idea, and it could be done for isometry groups too. Though, I would expect GAP already uses the methods of the determinant to evaluate membership in GL and SL; the contrary would be very strange.
2. That's why I was considering the idea that if `true` is returned by the line `G==H`, then the defined fields of `G` should be assigned to `H` and viceversa; just to add information and avoid useless computations. Anyway, the idea to use more direct membeship tests other than the GAP standard membership test is surely not optimal if many membership tests are made on the same group `G`; indeed, the "slow part" of the GAP membership test is the just computation of the GAP object and its composition tree. We can decide to adopt just the GAP membership test for everything and reduce the lines of the function `lies_in`.
3. At the moment, all the groups I'm interested in are already supported by GAP, all tho sometimes not in the way I like (e.g. the Omega groups are not subgroups of GO / SO, because they preserve different quadratic forms). The only exception is, of course, the groups with base_ring not supported by GAP.

ThomasBreuer replied:
Concerning the relations between GO, SO, and Omega in GAP, see the discussion at gap-system/gap/issues/500.
That is, GAP should provide the possibility to create the isometry group w.r.t. a given form.

GDeFranceschi replied:
Yeah, it is just a matter to compute a conjugate subgroup. I would like to define the standard copies of GO and SO to preserve the same form as Omega (that is the same of Magma).

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -223,6 +222,13 @@ function isquaternion_group(G::GAPGroup)
+
+
+
+####  This part is no longer active
```
So why not just remove that code?

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+#
+################################################################################
+
+function elem_f(x::fq_nmod; B=0,d=0)
```
what does this do? what does the name stand for? please add a comment

fingolfin replied:
Looking at the calls, B and d are always passed, so why do you make them keyword arguments (which don't participate in method dispatch)? I'd turn them into regular positional arguments, then you can also get rid of the nonsense default values.

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+   return x_gap
+end
+
+function elem_g(x::FFE; B=0, z=0, d=0)
```
Same remarks as for `elem_f` apply here, too

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+   return sum([L[i]*z^(i-1) for i in 1:d])
+end
+
+function gen_ring_iso(F::FqNmodFiniteField)
```
please add a comment explaining what this does (at least: what's the input, what's the output)

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+end
+
+function elem_g(x::FFE; B=0, z=0, d=0)
+   L = GAP.gap_to_julia(GAP.Globals.List(GAP.Globals.Coefficients(B,x),GAP.Globals.IntFFE ))
```
What's the output type? a vector of ints? Then I'd recommend something like this instead of using `gap_to_julia`:
```suggestion
   L = Vector{Int}(GAP.Globals.List(GAP.Globals.Coefficients(B,x),GAP.Globals.IntFFE ))
```

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+   z = gen(F)
+
+   if d==1
+      f(x::fq_nmod) = Int(coeff(x,0))*GAP.Globals.One(GAP.Globals.GF(p))
```
This will call `GAP.Globals.One(GAP.Globals.GF(p))` each time it is called. Better to store that value and use it in a closure. Here's some micro benchmark:
```julia
function HACK(p::Int)
  tmpval = GAP.Globals.One(GAP.Globals.GF(p))
  f(x) = x*GAP.Globals.One(GAP.Globals.GF(p))
  g(x) = x* tmpval
  return f, g
end
```
and then
```
julia> using BenchmarkTools

julia> f31, g31 = HACK(31)
(var"#f#3"{Int64}(31), var"#g#4"{FFE}(GAP: Z(31)^0))

julia> @btime f31(1)
  426.919 ns (5 allocations: 96 bytes)
GAP: Z(31)^0

julia> @btime g31(1)
  100.483 ns (2 allocations: 32 bytes)
GAP: Z(31)^0

```
So this is a factor 4 in performance in this example.

fingolfin replied:
This is still not resolved. Can you perhaps just try the following:
```suggestion
      myOne = GAP.Globals.One(GAP.Globals.GF(p))
      f(x::fq_nmod) = Int(coeff(x,0))*myOne
```
I.e., does that produce correct results? Can you compare it with the existing code and using `BenchmarkTools`  / `@btime` as I did above?

GDeFranceschi replied:
Done it. Running time and allocated memory get halved.

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+#
+################################################################################
+
+function mat_oscar_gap(x::fq_nmod_mat; n=0, r=0)
```
I really advise agains tall those keyword arguments, here and elsewhere.

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+   end
+
+   return GAP.julia_to_gap(S)
+end
```
This is not covered by the test suite according to https://app.codecov.io/gh/oscar-system/Oscar.jl/compare/173/diff; please add tests exercising this function (ideally both directly and indirectly; i.e. some tests calling this directly; and some tests which tests functions ultimately calling this).

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,103 @@
+   COEF = [Int64(coeff(x,i)) for i in 0:d-1]
+   x_gap = sum([COEF[i]*B[i] for i in 1:d])
+   return x_gap
+end
```
This function is not covered by the test suite according to https://app.codecov.io/gh/oscar-system/Oscar.jl/compare/173/diff; please add tests exercising this function (ideally both directly and indirectly; i.e. some tests calling this directly; and some tests which tests functions ultimately calling this).

GDeFranceschi replied:
Do I write tests also for `elem_f`? This is just an internal auxiliary function, it is not made to be used by users.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+   r = MatrixGroup(m,F)
+   r.gens = L
+   return r
+end
```
This is not tested according to https://app.codecov.io/gh/oscar-system/Oscar.jl/compare/173/diff#D4 ; please go through that report and add tests for those things which need them. But do this only after the other changes (e.g. no point in adding tests for code that is later removed)

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+mutable struct MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: AbstractMatrixGroupElem
+   parent::MatrixGroup{RE, T}
+   elm::T
+   X::GapObj
```
A comment explaining what the relationship is between elm and X would be helpful. In particular, this should explain whether both always be set; or whether it is OK to set just the one but not the other -- and also vice-versa? Or is one member always expected to be set?

This then serves as specification for the rest of the code. Without it, it becomes difficult to maintain this code, as the "contract" for using this type is unclear.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+function Base.getproperty(G::MatrixGroup, sym::Symbol)
+
+   if sym === :mat_iso || sym === :ring_iso
+      if isdefined(G,sym) return getfield(G,sym) end
```
I think this should be the first line of this function (and then you can remove the analog lines like `if isdefined(G,:X) return getfield(G,:X) end` below).

Overall, this trick seems to get a bit out of hand; I am worried once again that this might end up de-optimizing a lot of code for the sake of some special cases. But since it is e.g. unclear to me when `mat_iso` / `ring_iso` will be set and when not, and when we might have to compute them on the fly, it is difficult for me to evaluate this.

GDeFranceschi replied:
`mat_iso` and `ring_iso` are computed and set the first time the underlying GAP object is needed for the group `G` or one its element `x`. I would say, one does not need to keep track of `mat_iso` and `ring_iso`, since they are automatically set once `G.X` or `x.X` are used. Should I write this as a comment somewhere in the code?

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+      if isdefined(G,sym) return getfield(G,sym) end
+      fm = gen_mat_iso(G.deg, G.ring)
+      G.ring_iso=fm.fr
+      G.mat_iso=fm
```
I see no explicit `return` statement here. So I think it'll return the result of the last evaluated line, which means `fm`. Assuming this is correct, then when asking for `G.ring_iso` the first time, we might get a wrong result (namely the `mat_iso` instead of the ring iso). Can you test that theory?

One fix would be to simply end the function with `return getfield(G, sym)`

GDeFranceschi replied:
Good observation! I corrected it.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+   elseif G.descr==:SO G.X=GAP.Globals.SO(0,G.deg,G.ring_iso.codomain)
+   elseif G.descr==:Omega G.X=GAP.Globals.Omega(0,G.deg,Int(order(G.ring)))
+   elseif G.descr==:GU G.X=GAP.Globals.GU(G.deg,Int(characteristic(G.ring)^(div(degree(G.ring),2) ) ))
+   elseif G.descr==:SU G.X=GAP.Globals.SU(G.deg,Int(characteristic(G.ring)^(div(degree(G.ring),2) ) ))
```
```suggestion
   elseif G.descr==:SU G.X=GAP.Globals.SU(G.deg,Int(characteristic(G.ring)^(div(degree(G.ring),2) ) ))
   else error("unsupported description")
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+end
+
+
+# add field :X
```
please update and expand this comment; also add a similar comment before the other `getproperty` method. At least one of them should really explain what's going on here.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+      return G.descr == H.descr
+   end
+   if isdefined(G, :gens) && isdefined(H, :gens)
+      if G.gens==H.gens return true  end       #TODO this works only if the == between elements does not keep track of the parent
```
What's to be done, though? If the parent is tracked, then the comparison will fail -- and we run into the generic case below. So no harm done. But if the parent *is* tracked... well, so what can we do? Nothing, I think? So what's the point of having this `TODO` comment -- perhaps you *do* have an idea how to handle this better??

GDeFranceschi replied:
At the moment the comparison `==` between matrix group elements just checks if the matrices coincide, and does not check also whether the parents are the same (for example, `one(GL(n,q))==one(Sp(n,q))` would return `true`). This is also how Magma acts, and this is also how Oscar does for other groups (for example, `one(symmetric_group(5))==one(alternating_group(5))` returns also `true`).
If we decide in future to allow `x==y` only for `parent(x)==parent(y)`, then checking if two matrix groups coincide by checking their generating sets does no longer work, because an infinite loop would be created.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+end
+
+# embedding a nxn array into a group G
+function (G::MatrixGroup)(L::AbstractVector)
```
I don't understand -- an n x n array would be a Matrix, not a Vector. Comment and code don't seem to match? also, what's the usecase for this function?

GDeFranceschi replied:
Yes, it's a matrix. It's just a help, so if one wants to write an explicit matrix in the group `G`, he can just write more concisely `G([a,b,c,d])` instead of `G(matrix(F,2,2,[a,b,c,d]))`.

rfourquet replied:
> he can just write more concisely `G([a,b,c,d])` instead of `G(matrix(F,2,2,[a,b,c,d]))`.

Is it meant mostly for literal arrays? Then maybe it would be useful (instead or in addition to) to allow passing an `AbstractMatrix`, which can be entered as succintly as a vector (`G[a b; c d]`). (I personally prefer the `AbstractMatrix` version, which is more explicit, in particular regarding row vs. column major order). If you want to keep both, you could relax `L::AbstractVector` to `L::AbstractVecOrMat`.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+#
+########################################################################
+
+# TODO: we are not currently keeping track of the parent
```
huh, but isn't there a `parent` field? I am confused.

GDeFranceschi replied:
Two variables of type `MatrixGroupElem` coincide if, and only if, their matrices coincide, regardless of their parent. E.g. `one(GL(n,q))==one(SL(n,q))` returns `true`.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+   end
+end
+
+# TODO: we wish to multiply / conjugate also matrices with different parents ?
```
Largely depends on how you defined parents... The rules for this should be written down somewhere, then we can decide this.

GDeFranceschi replied:
A situation occurring quite often where we multiply matrices with different parents is when one of the parents is a subgroup of the other one; in such a case, the ideal output parent would be the bigger group, but this could lead to many useless and expensive `issubgroup` and membership tests.
Anyway, I added a parameter `check` to the embedding functions. The user can disable this check if he is sure of the membership and skip all the checks.
A possibility is to set the parent of a product as simply `GL(n,q)`.

GDeFranceschi replied:
Magma, for example, sets the parent of `x*y` as `GL(n,q)` whenever `x` and `y` have different parents (but same degree and base ring, of course). This is for sure the easiest way to proceed and avoid plenty of useless checks and membership tests.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+
+########################################################################
+#
+# conversions     (this should be no longer necessary, and it's also wrong)
```
then please remove this code (we can always recover it from the git history)

thofma commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,940 @@
+   mat_iso::GenMatIso
+   AbstractAlgebra.@declare_other
+
+   MatrixGroup(m::Int, F::Ring) = new{elem_type(F), MatElem{elem_type(F)}}(m,F)
```
```suggestion
   MatrixGroup(m::Int, F::Ring) = new{elem_type(F), dense_matrix_type(elem_type(F))}(m,F)
```

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/iso_oscar_gap.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: thofma <thofma@gmail.com>

GDeFranceschi committed:
Suggestions from PR #173

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

ThomasBreuer commented:
@GDeFranceschi Currently something seems to be broken.
When I create `g = general_linear_group(2,2)` then `one(g)`, `g.gens` run into errors.

GDeFranceschi committed:
Suggestions from PR #173

Commit:
Little change + tests

GDeFranceschi committed:
Little changes

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+end
+
+# build a MatrixGroup given a list of generators
+function MatrixGroup(m::Int, F::S, L::Vector) where S<:Ring
```
```suggestion
function MatrixGroup(m::Int, F::Ring, L::Vector)
```
In addition, shouldn't the type of `L` be restricted further e.g. to `Vector{<:AbstractMatrixGroupElem}` ?

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+# build a MatrixGroup given the underlying GAP object
+function MatrixGroup(m::Int, F::S, G_gap::GapObj) where S<:Ring
+   r = MatrixGroup(m,F)
+   r.X = G_gap
```
This constructor is unsafe to use, as it does not ensure that `G_gap` is a GAP matrix group, nor that the dimension/degree and base ring match `m` resp. `F`. (BTW I suggest renaming `F` to `R`).

GDeFranceschi replied:
I just remove that constructor. I just realized that it creates a `MatrixGroupElem` object where the field `:X` is defined and the fields `:mat_iso` and `:ring_iso` are not.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+function MatrixGroup(m::Int, F::S, L::Vector) where S<:Ring
+   r = MatrixGroup(m,F)
+   r.gens = L
+   for l in r.gens l.parent = r end
```
Doesn't this modify the argument? That's something one should avoid in general, as it can cause endless bugs. Better to make a copy of the inputs then? Also, how about a check that the dimensions (and possibly the base ring) match?

GDeFranceschi replied:
I have just rewritten it.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+end
+
+
+# NOTE: at least one of the fields :elm and :X is always defined, but not necessarly both of them.
```
```suggestion
# NOTE: at least one of the fields :elm and :X must always defined, but not necessarily both of them.
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+
+# build a MatrixGroupElem given its underlying GAP object
+# WARNING: this does not check whether the element actually lies in the group G
+function MatrixGroupElem(G::MatrixGroup, x::T, x_gap::GapObj)  where T <: MatElem
```
```suggestion
function MatrixGroupElem(G::MatrixGroup, x::MatElem, x_gap::GapObj)
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+end
+
+# build a MatrixGroupElem given its underlying GAP object
+# WARNING: this does not check whether the element actually lies in the group G
```
Hmm.... shouldn't it be rather be something like this?
```suggestion
# build a MatrixGroupElem given both an OSCAR matrix and a corresponding GAP object.
# WARNING: this does not check whether the element actually lies in the group G
# nor whether the two given matrices really match.
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+
+At the moment, only rings of type `FqNmodFiniteField` are supported.
+"""
+mutable struct MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
```
A comment on the meta level (i.e., I don't require you to address this in this PR; but it may have to be addressed at some point in the future anyway, during a major refactoring): I don't like that we are effectively "hardcoding" the matrix groups must be implemented via underlying GAP objects. To an extent, that's the fault of Julia's type system which makes it difficult to makes these kinds of things difficult; but I still feel we could do better. 

On the other hand, providing actual alternate group types which are not based on GAP code (but e.g. instead on pure Julia code) but which still provide a similar level of functionality would be really challenging (I have some ideas on that, but let's not get distracted by that right now). The question is: why would we do that? What benefit do we expect? I am concerned that without a clear answer to that, any attempt to "generalize" this is doomed to be (a) insufficient, and (b) a waste of resources... Yet at the same time I can't help but feel unsatisfied with this :-/.

fingolfin reviewed (commented):
Some more comments...

@GDeFranceschi do you think you could rebase this on the current Oscar.jl master branch, so that we can take advantages in GAP.jl 0.5.0? I could possibly help you with that tomorrow (Thursday), e.g. via screen sharing

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,854 @@
+function Base.show(io::IO, x::MatrixGroupElem)          #TODO: is this correct?
+   if isdefined(x, :elm)
+      show(io, "text/plain", x.elm)
+#      print(io, x.elm)
```
To quote the documentation for `Base.show` on two arguments:
> The representation used by show generally includes Julia-specific formatting and type information, and should be parseable Julia code when possible.
> ...
> To customize human-readable text output for objects of type T, define show(io::IO, ::MIME"text/plain", ::T) instead.

So perhaps we should *not* do this, but instead call just plain `show(io, x.elm)`, and then add another method `Base.show(::IO, ::MIME"text/plain", : x::MatrixGroupElem)` which invokes `show(io, "text/plain", x.elm)`???

Perhaps @rfourquet has some insight into this?

rfourquet replied:
the `show` machinery is known to be difficult to understand in Julia... I would agree with @fingolfin here for not calling the 3-args `show` method within the 2-args one. And I don't know of a (simple) way of avoiding writing two different `show` methods (2- and 3-args) which just forward the presence or absence of a `MIME` parameter.

thofma replied:
The printing/show machinery is indeed difficult. At the moment we have the following for matrices:
```julia
julia> A = matrix(ZZ, 2, 2, [1, 2, 3, 4])
[1  2]
[3  4]

julia> [A, A]
2-element Array{AbstractAlgebra.Generic.MatSpaceElem{BigInt},1}:
 [1 2; 3 4]
 [1 2; 3 4]

julia> print(A)
[1 2; 3 4]
```
I guess we want to the same behavior? The "compact" behavior is coming from overloading
```julia
function Base.show(io::IO, a::T)
```
The ordinary REPL printing comes from overloading
```julia
function Base.show(io::IO, ::MIME"text/plain", a::T)
```

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Suggestions from PR #173

GDeFranceschi committed:
Suggestions from PR #173 + Tests

GDeFranceschi committed:
Little changes + tests

GDeFranceschi committed:
Add field order to Matrix type

fingolfin commented:
@GDeFranceschi please take the comments I made on PR #244 and apply them here.

Then, you really should focus on making the tests pass! We won't be able to merge anything if it breaks the tests. (One of the reasons I didn't comment more on your PRs was/is that I was waiting for you to stabilize the tests... please make this a priority)

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,892 @@
+      z = new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}()
+      z.parent = G
+      z.X = x_gap
+      return z
```
If you swap the definitions of `elm` and `X` (making `elm` the last member of the struct), then I *think* you can change this to
```suggestion
      return new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}(G, x_gap)
```

GDeFranceschi replied:
Yes, but in this case I have to write a longer version for the other constructor (the one that defines `parent` and `elm`, but not `X`). At least one of them need to be long.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -63,10 +63,7 @@ export
-function group_element(G::T, x::GapObj) where T <: GAPGroup
-  return GAPGroupElem{T}(G, x)
-end
+group_element(G::T, x::GapObj) where T <: GAPGroup = BasicGAPGroupElem{T}(G,x)
```
Perhaps you can extract the change(s) which introduce `BasicGAPGroupElem` into a standalone PR, which then would be fairly small and could be merged quickly; that way, this PR here would not touch quite as many files.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,892 @@
+trace(x::MatrixGroupElem) = trace(x.elm)
+tr(x::MatrixGroupElem) = tr(x.elm)
+
+order(x::MatrixGroupElem) = GAP.Globals.Order(x.X)
```
I commented on this in another PR, suggesting to turn this into
```suggestion
order(x::MatrixGroupElem) = ZZ(GAP.Globals.Order(x.X))
```
but actually to match the generic `order` for GAP groups, we probably should also have this:
```suggestion
order(x::MatrixGroupElem) = ZZ(GAP.Globals.Order(x.X))
order(::Type{T}, x:: MatrixGroupElem) = T(GAP.Globals.Order(x.X)) where {T}
```

GDeFranceschi committed:
Suggestions from PR#173 and PR#244

GDeFranceschi committed:
rebase on BasicGAPGroupElem

fingolfin commented on src/Groups/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,113 @@
+################################################################################
+
+# return the GAP matrix corresponding to the Oscar matrix x
+function mat_oscar_gap(x::fq_nmod_mat, n, r)
```
what is `n`? I assume an integer specifying the dimension of the output matrix. Why is this not just taken from `x`? And what is `r`? A ring, or a conversion function, or what? Looking at callers, there it is called `riso`, indicating it is an isomorphism -- I suggest to use this name here (and in similar places) for clarity.
```suggestion
function mat_oscar_gap(x::fq_nmod_mat, n::Int, r)
```

GDeFranceschi replied:
I just removed the variable `n`. Now, `mat_oscar_gap` works on matrices of every dimensions.

fingolfin commented on src/Groups/types.jl:
```diff
@@ -77,6 +75,7 @@ TODO: document this
 abstract type GAPGroup <: AbstractAlgebra.Group end
 abstract type GAPGroupElem{T<:GAPGroup} <: AbstractAlgebra.GroupElem end
 
+abstract type GAPGroupElem{T<:GAPGroup} <: AbstractAlgebra.GroupElem end
```
```suggestion
```
I think this line is redundant (it is identical to the one above?). Perhaps this slipped in during a rebase?

GDeFranceschi replied:
Yes, probably.

fingolfin commented on test/Groups/matrixgroups.jl:
```diff
@@ -0,0 +1,421 @@
+   for i in 1:3
+      xg[i] = GAP.julia_to_gap([G.ring_iso(xo[i,j]) for j in 1:3])
+   end
+   xg=GAP.julia_to_gap(xg)
```
I recommend to avoid `GAP.julia_to_gap` and instead use `GapObj`:
```suggestion
   xg = Vector{GapObj}(undef, 3)
   for i in 1:3
      xg[i] = GapObj([G.ring_iso(xo[i,j]) for j in 1:3])
   end
   xg=GAP.julia_to_gap(xg)
```

But perhaps even better would be to use a recursive conversion (this is *untested*, so don't blindly merge this suggestion)
```suggestion
   xg=GapObj([G.ring_iso(xo[i,j]) for j in 1:3 for j in 1:3]; recursive = true)
```

GDeFranceschi replied:
The line `xg = GapObj([[G.ring_iso(xo[i,j]) for j in 1:3] for i in 1:3]; recursive=true)` works and passes the tests. Do I leave it?

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -386,22 +386,27 @@ function hom(x::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup
+    @assert A.G == G || GAP.Globals.IN(x.X, A.G.X) "Not in the domain of f!"      #TODO Do we really need the IN check?
+  end
+  if x isa MatrixGroupElem
+     return MatrixGroupElem(G,GAP.Globals.Image(f.X,x.X))
```
I don't understand why this is necessary, can you explain? It is also not covered by a test -- perhaps you can write a code example which triggers this, and then tell us which error this code example produces if one removes this special case.

GDeFranceschi replied:
Do you mean the assertion or the `if isa MatrixGroupElem` ?
The first one is to avoid to compute the automorphism on an element that is not in its domain, that creates a long GAP error.
For example: the following piece of code

> A = automorphism_group( alternating_group(5)); f = rand(A);
> x = cperm([1,3,5,4]);
> Oscar.apply_automorphism(f,x, false);

produces the error
> ERROR: Error thrown by GAP: Error, <elm> must be an element of Source(<map>)

The distinction between `MatrixGroupElem` and the other types is that if `x` is a matrix, then its type is `MatrixGroupElem{T}`, but the constructor is simply `MatrixGroupElem` (without the parameter), so the line
> return typeof(x)(G, GAP.Globals.Image(f.X,x.X))

does not work.

fingolfin replied:
That's arguably a bug (or misfeature) in your code: there *should* be a `MatrixGroupElem{T}` constructor. Please take a look at https://github.com/oscar-system/Oscar.jl/pull/267 where I resolved a very similar issue

GDeFranceschi replied:
I wrote the parametric constructors for `MatrixGroup` and `MatrixGroupElem`. What if the arguments of the constructors have a different type from the parameters? Or suptype/subtype? I mean, what should happen if I have something like `MatrixGroupElem{RE, T}(G, x::S)`, where the `S` is a subtype/suptype of `T` ? Do I return a variable of type `MatrixGroupElem{RE,S}`? Or a variable of type `MatrixGroupElem{RE,T}`? Or an ERROR message?
For the moment I lean towards the first option. Is it fine? Or are there better solutions?

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -386,22 +386,27 @@ function hom(x::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup
-Base.:^(x::GAPGroupElem{T},f::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup = apply_automorphism(f, x)
+(f::GAPGroupElem{AutomorphismGroup{T}})(x::GAPGroupElem) where T <: GAPGroup = apply_automorphism(f, x)
+Base.:^(x::GAPGroupElem,f::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup = apply_automorphism(f, x)
+Base.:^(f::GAPGroupElem{AutomorphismGroup{T}},g::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup = g^-1*f*g
```
I think the only reason you need this line above is because our generic definition for conjugation is unnecessarily ambiguous. It currently says this:
```julia
Base.:^(x::GAPGroupElem, y::GAPGroupElem) = ...
```
I think if you change that as follows, the above line may not be needed:
```julia
Base.:^(x::T, y::T) where {T <: GAPGroupElem} = ...
```

GDeFranceschi replied:
It's needed anyway, because the operation `x^f`, intended as the evaluation of the automorphism `f` in the variable `x`, isn't defined anywhere else.

fingolfin replied:
Sure, that's the one in the line *above*, i.e., `Base.:^(x::GAPGroupElem,f::GAPGroupElem{AutomorphismGroup{T}})`.

But I am talking about `Base.:^(f::GAPGroupElem{AutomorphismGroup{T}},g::GAPGroupElem{AutomorphismGroup{T}})`

fingolfin replied:
Ping?

GDeFranceschi replied:
Done it.

GDeFranceschi committed:
Tests + suggestions from PR #173

Commit:
More tests

GDeFranceschi committed:
Update runtests.jl

GDeFranceschi committed:
Update runtests.jl

GDeFranceschi committed:
Adjustments in tests

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+Base.in(x::MatElem, G::MatrixGroup) = lies_in(x,G,Nothing)[1]
+
+function Base.in(x::MatrixGroupElem, G::MatrixGroup)
+   isdefined(x,:X) && lies_in(x.elm,G,x.X)[1]
```
Do you perhaps means this? Otherwise I don't understand the code at all
```suggestion
   isdefined(x,:X) && return lies_in(x.elm,G,x.X)[1]
```

GDeFranceschi replied:
Yes, of course :-)

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+   isdefined(x,:X) && lies_in(x.elm,G,x.X)[1]
+   vero, x_gap = lies_in(x.elm,G,Nothing)
+   if x_gap !=Nothing x.X = x_gap end
+   return vero
```
I think you want to use the value `nothing`, not the type `Nothing`:
```suggestion
   vero, x_gap = lies_in(x.elm,G,nothing)
   if x_gap !== nothing x.X = x_gap end
   return vero
```

fingolfin replied:
Also, what does `vero` mean? Is this perhaps Italian?

GDeFranceschi replied:
Yes, "vero" means "true" in Italian, but I can't use a variable named "true".

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+      end
+   else
+      return MatrixGroupElem(G,x)
+   end
```
```suggestion
function (G::MatrixGroup)(x::MatElem; check=true)
   if check
      vero, x_gap = lies_in(x,G,nothing)
      vero || throw(ArgumentError("Element not in the group"))
      if x_gap !== nothing
         return MatrixGroupElem(G,x,x_gap)
      end
   end
   return MatrixGroupElem(G,x)
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+   
+   L = GAP.Globals.GeneratorsOfGroup(G.X)
+   G.gens=[MatrixGroupElem(G,G.mat_iso(L[i]),L[i]) for i in 1:length(L)]
+   return G.gens
```
```suggestion
   if !isdefined(G,:gens)
      L = GAP.Globals.GeneratorsOfGroup(G.X)
      G.gens=[MatrixGroupElem(G,G.mat_iso(a),a) for a in L]
   end
   return G.gens
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+gen(G::MatrixGroup, i::Int) = gens(G)[i]
+
+Base.getindex(G::MatrixGroup, i::Int) = gen(G, i)
+
```
Should be superfluous as it is identical to the `getindex` method for `GAPGroup`
```suggestion
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+
+   G.order = fmpz(BigInt(GAP.Globals.Size(G.X)))
+   return G.order
+end
```
This breaks the contract for `order` as it doesn't implement the two argument version. In fact, with my PR #269 it will be sufficient to only implement the two argument version here, say like this:
```suggestion
function order(::Type{T}, G::MatrixGroup) where T
   isdefined(G, :order) || G.order = fmpz(GAP.Globals.Order(G.X))
   return T(G.order)
end
```
However, I still am not convinced that having the `G.order` field is a good idea in the first place; if we drop it, we can also drop this method and just rely on the generic one.

GDeFranceschi replied:
It was Claus' suggestion. If I do not have the field `G.order`, I don't have way to store the order other than the underlying GAP object.

GDeFranceschi replied:
The line `isdefined(G, :order) || G.order = fmpz(GAP.Globals.Order(G.X))` returns the error `ERROR: syntax: invalid assignment location`. I replaced it by `if !isdefined(G, :order) G.order =  fmpz(GAP.Globals.Order(G.X)) end`.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+
+function general_linear_group(n::Int, q::Int)
+   (a,b) = ispower(q)
+   if !isprime(b) throw(ArgumentError("The field size must be a prime power")) end
```
More idiomatic:
```suggestion
   isprime(b) || throw(ArgumentError("The field size must be a prime power"))
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+   elseif T<:MatrixGroup   return general_linear_group(n,F)
+   else throw(ArgumentError("Type not supported"))
+   end
+end
```
Are these really useful? I don't recall ever having used `GL(IsPermGroup, n, q)` in GAP (if at all, I'd have used the image of `IsomorphismPermGroup(GL(n,q,))`.

Thing is, the methods as you wrote them above have serious issues, in that they accept crazy stuff like `general_linear_group(MatrixGroup{Integers}, 3, 3)`. While one could fix that (e.g. by using `T <: MatrixGroup && return T(general_linear_group(n,F))`) I am wondering whether it is worth the effort?

This is a good example for my rule of thumb, "if you have no usecase for code, don't write it, it'll be wrong"

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+   else throw(ArgumentError("Type not supported"))
+   end
+end
+
```
See above: just discard these? I won't repeat this again and again, though.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+tr(x::MatrixGroupElem) = tr(x.elm)
+
+order(x::MatrixGroupElem) = ZZ(GAP.Globals.Order(x.X))
+order(::Type{T}, x::MatrixGroupElem)  where T<:Number = T(GAP.Globals.Order(x.X))
```
```suggestion
order(x::MatrixGroupElem) = order(fmpz, x)
order(::Type{T}, x::MatrixGroupElem)  where T = T(GAP.Globals.Order(x.X))
```
However, we may also be able to just remove these two lines if PR #269 is merged?!

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+   descr::Symbol                       # e.g. GL, SL, symbols for isometry groups
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+   order::fmpz
```
Where is this used or set?

GDeFranceschi replied:
`:descr` is set during the construction, and it is used in the `show` method and in the setting of `:X`.
`:ring_iso` and `:mat_iso` are set when the field `:X` is set, or when the field `:X` of any element of `G` is set.

fingolfin replied:
OK, sure, but I was specifically wondering about the line I commented on: what is `order` needed for?

GDeFranceschi replied:
It serves just to store the value (similarly to what happens in GAP objects) and avoid expensive re-computations. I introduced it for the cases the order can be computed without using GAP procedures (there are not examples in this branch, but there are, for example, in the branch `jordan-struct`, where the order of centralizers in SL and GL are computed independently).
The alternative would be to set the field `:X` (the underlying GAP object) whenever call the function `order` and using `SetSize` on the GAP object `G.X`.

fingolfin replied:
But that's what we have `@declare_other` and the code related to it for: to store optional data that only some instances of this type want to store.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,888 @@
+function sub(G::MatrixGroup, elements::Vector{S}) where S <: GAPGroupElem
+   @assert elem_type(G) == S
+   elems_in_GAP = GAP.julia_to_gap(GapObj[x.X for x in elements])
+   H = GAP.Globals.Group(elems_in_GAP)
```
Probably better to have this tracked as a subgroup in GAP, too (so it can use knowledge from the parent group)... 
```suggestion
   H = GAP.Globals.Subgroup(G.X, elems_in_GAP)
```
The same should be fixed in the general `sub` for GAPGroup.

Except that perhaps `G.X` might not be defined?!? Perhaps need a separate case for that?

GDeFranceschi replied:
For the moment, we decided that group homomorphisms always have an underlying GAP object, and that the methods `sub(G, [...])` always returns the subgroup `H` together with the inclusion homomorphism of `H` into `G`. Hence, every time we compute the subgroup of a matrix group `G`, the underlying GAP object is always computed as well.

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -213,21 +193,21 @@ function _as_subgroup(G::DirectProductGroup, H::GapObj)
-function sub(G::DirectProductGroup, elms::Vector{GAPGroupElem{DirectProductGroup}})
+function sub(G::DirectProductGroup, elms::Vector{<:GAPGroupElem{DirectProductGroup}})
   elems_in_GAP = GAP.julia_to_gap(GapObj[x.X for x in elms])
   H = GAP.Globals.Group(elems_in_GAP)
```
```suggestion
  H = GAP.Globals.Subgroup(G.x, elems_in_GAP)
```

fingolfin commented on src/Groups/directproducts.jl:
```diff
@@ -371,21 +351,21 @@ function _as_subgroup(G::SemidirectProductGroup{S,T}, H::GapObj) where S <: GAPG
-function sub(G::SemidirectProductGroup{S,T}, elms::Vector{GAPGroupElem{SemidirectProductGroup{S,T}}}) where { S, T }
+function sub(G::SemidirectProductGroup{S,T}, elms::Vector{<:GAPGroupElem{SemidirectProductGroup{S,T}}}) where { S, T }
   elems_in_GAP = GAP.julia_to_gap(GapObj[x.X for x in elms])
   H = GAP.Globals.Group(elems_in_GAP)
```
see remark above (should be fixed in all affected places)

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -386,22 +386,27 @@ function hom(x::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup
+#     return MatrixGroupElem(G,GAP.Globals.Image(f.X,x.X))
+#  else
+     return typeof(x)(G, GAP.Globals.Image(f.X,x.X))
+ # end
```
```suggestion
  return typeof(x)(G, GAP.Globals.Image(f.X,x.X))
```

GDeFranceschi replied:
The line `Base.:^(f::GAPGroupElem{AutomorphismGroup{T}},g::GAPGroupElem{AutomorphismGroup{T}})` defines just the conjugation operation in automorphism groups, that is defined as in the other types of groups. Removing that line would resolve in an error when typing `f^g`, because Oscar does not recognize `f` as an element in `domain(g)`.

GDeFranceschi committed:
Suggestions from PR #173

GDeFranceschi committed:
Further from PR #173

GDeFranceschi committed:
Remove order(G::MatrixGroup) + some tests

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,790 @@
+   X::GapObj
+   gens::Vector{<:AbstractMatrixGroupElem}
+   descr::Symbol                       # e.g. GL, SL, symbols for isometry groups
+   ring_iso::GenRingIso
```
It is actually unclear to me why `ring_iso` is needed -- aren't you only every using `mat_iso`? I had a look at the code and my impression was that the only place using `ring_iso` is `assign_from_description`, and that only uses it to get `G.ring_iso.codomain`.

Ah, and also `@testset "Oscar-GAP relationship" begin` accesses it -- but I think that one could (and should) just be rewritten to test `mat_iso` instead (which is not tested there at all right now, as far as I could tell?); and I don't just mean that one should replace `G.ring_iso` by `G.mat_iso.fr`, but rather that the `mat_iso` function should be tested by mapping matrices back and forth, as that is the workhorse of the conversion, isn't it?

GDeFranceschi replied:
Done it. I just removed the field `:ring_iso`.

Commit:
Remove :ring_iso + parametric constructors

GDeFranceschi committed:
Fix parametric constructors

fingolfin commented:
Can you also rebase this?

GDeFranceschi committed:
Remove omega_group(-1,2,q) + change Nothing -> nothing

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,790 @@
+
+end
+
+MatrixGroup{RE,T}(m::Int, F::Ring) where {RE,T} = MatrixGroup(m,F)
```
This is backwards: the parametric constructor should be "inner" and the default, and the non-parametric one should call it. With your construction, the parameters are ignored!

GDeFranceschi replied:
The line
> MatrixGroup{RE,T}(m::Int, F::Ring) where {RE,T} = new{RE,T}(m,F)

defines the default constructor. Should I add a check to guarantee the parametric type is correct? This is something like
> @ assert element_type(F) <: RE

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,790 @@
+   return G
+end
+
+MatrixGroup{RE,S}(m::Int, F::Ring, V::AbstractVector{T}) where {RE,S} where T<:Union{MatElem,AbstractMatrixGroupElem} = MatrixGroup(m,F,V)
```
See above

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,790 @@
+         else
+            L[i] = MatrixGroupElem(G,V[i].X)
+         end
+# once deepcopy works for GAP objects
```
Maybe this line should be moved upwards, to lines 61/62? I was really confused by lines 61/62 (as in: "so why not do it?") until I saw this line

GDeFranceschi committed:
Switch parametric constructors as default@

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+   z = MatrixGroupElem{RE,T}(G,x)
+   z.X = x_gap
+   return z
+end
```
This makes no sense as written. How about this
```suggestion
   # full constructor
   MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE, T} = new{RE,T}(G, x, x_gap)

   # constructor which leaves `X` undefined
   MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x::T) where {RE, T} = new{RE,T}(G, x)

   # constructor which leaves `elm` undefined
   function MatrixGroupElem{RE,T}(G::MatrixGroup{RE,T}, x_gap::GapObj) where {RE, T}
      z = new{RE,T}(G)
      z.X = x_gap
      return z
   end

end

MatrixGroupElem(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE,T} = MatrixGroupElem{RE,T}(G, x, x_gap)
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+
+function Base.getproperty(x::MatrixGroupElem, sym::Symbol)
+
+   if isdefined(x,sym) return getfield(x,sym) end
```
For consistency with the preceding function...
```suggestion
   isdefined(x,sym) && return getfield(x,sym)
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+  return MatrixGroupElem(G, G.mat_iso(i),i), L
+end
+
+function Base.iterate(G::MatrixGroup, state)
```
```suggestion
function Base.iterate(G::MatrixGroup, state::GapObj)
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+         return true, x_gap
+      else
+         return false, x_gap
+      end
```
```suggestion
      return det(x)!=0, x_gap
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+         return true, x_gap
+      else
+         return false, x_gap
+      end
```
```suggestion
      return det(x)==1, x_gap
```

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+
+function Base.in(x::MatrixGroupElem, G::MatrixGroup)
+   isdefined(x,:X) && return lies_in(x.elm,G,x.X)[1]
+   vero, x_gap = lies_in(x.elm,G,nothing)
```
`vero` is presumably Italian? Change it to english, please, here and elsewhere

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+   if !check
+      z = x
+      z.parent = G
+      return z
```
Untested code path, please add a test.

fingolfin commented on src/Groups/MatGrp.jl:
```diff
@@ -0,0 +1,792 @@
+         return MatrixGroupElem(G,x.elm,x.X)
+      else
+         GAP.Globals.in(x.X, G.X) || throw(ArgumentError("Element not in the group"))
+         return MatrixGroupElem(G,x.X)
```
Untested code path, please add a test.

fingolfin commented on src/Groups/sub.jl:
```diff
@@ -63,11 +67,11 @@ end
+Return (`true`,`f`) if `H` is a subgroup of `G`, where `f` is the embedding homomorphism of `H` into `G`, otherwise return (`false`,`nothing`).
 """
 function issubgroup(G::T, H::T) where T <: GAPGroup
    if false in [h in G for h in gens(H)]
```
I realize this code was not added in this PR, but let me point out that this is very inefficient, esp. if the number of generators is high: it has to create a vector of size equal to the number of generators, always looks at all generators, and has to search the whole generated vector; but in reality, a single bit as storage is sufficient, and we can abort immediately as soon as we find an element not in `G`.

Better alternative (and also more readable, IMHO)
```suggestion
   if !all(h -> h in G, gens(H))
```

fingolfin reviewed (commented):
A final round of comments. Let's merge it once those are addressed (or even if not), so that your other PRs can be hopefully taken care of, too

fingolfin commented on test/Groups/matrixgroups.jl:
```diff
@@ -0,0 +1,457 @@
+#      xg[i] = GapObj([G.mat_iso.fr(xo[i,j]) for j in 1:3])
+#   end
+#   xg=GAP.julia_to_gap(xg)
+   riso=Oscar.gen_ring_iso(F)
```
As far as I can tell, you "only" test if `riso` is bijective -- but not that it is a homomorphism. I would feel much better if there were a few simple tests. I.e., for finite fields of order 4, 8, 9, 25, 27,  you could simply iterate over all pairs a,b and compare the image of `a*b` with the product of the images of `a` and `b`.

GDeFranceschi replied:
Done it.

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/MatGrp.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Update src/Groups/sub.jl

Co-authored-by: Max Horn <max@quendi.de>

GDeFranceschi committed:
Some suggestions from PR #173

GDeFranceschi committed:
Remove field :order

GDeFranceschi committed:
Two more tests

GDeFranceschi committed:
More efficient gen_mat_iso + remove sub in directproducts.jl

Commit:
Put ImmutableMatrix in iso_oscar_gap.jl
